calyx-py/test/correctness/queues/rr_queues/rr_queue.py

```diff
@@ -0,0 +1,222 @@
+       handles.append(handle)
+
+   flow_handles = []
+   for b in range(n_flows):
+    #    _ , val_lt_cell = pifo.le_store_in_reg(value, boundaries[b+1])
+    #    _ , val_gt_cell = pifo.ge_store_in_reg(value, boundaries[b])
+    #    #val_add = pifo.add_store_in_reg(val_lt_cell.out, val_gt_cell.out, val_add_cell)
+    #    _ , flow_reg = pifo.eq_store_in_reg(val_lt_cell.out, val_gt_cell.out)
+       handle1 = cb.if_with(pifo.le_use(value, boundaries[b+1]), 
+       cb.if_with(pifo.ge_use(value, boundaries[b]), invoke_subqueue(fifo_cells[b], cmd, value, ans, err)))
+
+       flow_handles.append(handle1)
+
+   # The main logic.
+   pifo.control += cb.par(
+       # Was it a pop, peek, or a push? We can do all cases in parallel.
+       cb.if_with(
+           # Did the user call pop?
+           cmd_eq_0,
+           cb.if_with( len_eq_0, raise_err,  # The queue is empty: underflow.
+               [  # The queue is not empty. Proceed.
+                   lower_err,
+                   [ handles,
+                       # Our next step depends on whether `fifos[hot]` raised the error flag.
+                       cb.while_with( err_neq_0,
+                           [  # `fifo_cells[hot]` raised an error.
+                               # We'll try to pop from `fifo_cells[hot+1]`.
+                               # We'll pass it a lowered err
+                               lower_err,
+                               cb.if_with(hot_eq_n, reset_hot, incr_hot),
+                               handles, #TODO handles is the problem?!?!
+                           ], # `queue[hot+n]` succeeded. Its answer is our answer.
+                       ),                             
+                   ],
+                   cb.if_with(hot_eq_n, reset_hot, incr_hot), # are reset_hot anf incr_hot in the right order?
+                   len_decr, 
+               ],
+           ),
+       ),
+       cb.if_with(
+           # Did the user call peek?
+           cmd_eq_1,
+           cb.if_with( len_eq_0, raise_err,  # The queue is empty: underflow.
+               [  # The queue is not empty. Proceed.
+                   lower_err,
+                   copy_hot,
+                   [ handles,
+                       cb.while_with(
+                           err_neq_0,
+                           [  # `fifo_cells[hot]` raised an error.
+                               # We'll try to peek from `fifo_cells[hot+1]`.
+                               # We'll pass it a lowered `err`.
+                               lower_err,
+                               cb.if_with(hot_eq_n, reset_hot, incr_hot), # increment hot and invoke_subqueue on the next one
+                               handles
+                           ],
+                       ),
+                       # Peeking does not affect `hot`.
+                       # Peeking does not affect the length.
+                   ], 
+                   restore_hot,                
+               ],
+           ),
```

Comment:
Nice! Minor thing to make your life easier: a new-ish thing we have begun to do is to actually create a little Pytohn handle for _all_ this logic, just called `peek_logic` for example. Then the top-level control could just become: `cb.if_with(cmd_eq_1, peek_logic)`. 

Then, we have a brand new eDSL constructor for cases. That allows us to write even more readable code. Please see the below for a taste! https://github.com/calyxir/calyx/blob/ba95ac0b1159954e5c50472e028c043c1cb555c5/calyx-py/test/correctness/queues/fifo.py#L87-L95